File: stocvest/signals/setup_judgment.py
"""
Setup judgment — quality vs tradeability split (engine-facing + API).

Quality reflects multi-layer agreement (not a reweighted technical score).
Tradeability reflects entry geometry (extension, phase) — not R/R gates or clock time.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

from stocvest.analytics.unlock_forecast import (
    _layer_aligned,
    composite_bias_from_summary,
    derive_missing_layers,
)
from stocvest.models.watchlist import MATURATION_LAYER_KEYS

SetupPhase = Literal["early", "expansion", "extended", "exhaustion"]
ProcessTier = Literal["not_aligned", "developing", "near_ready", "actionable"]
TradeabilityBand = Literal["strong", "moderate", "weak"]
FlagSeverity = Literal["warn", "block"]

_LAYER_LABEL: dict[str, str] = {
    "technical": "Technical",
    "news": "News",
    "macro": "Macro",
    "sector": "Sector",
    "geopolitical": "Geopolitical",
    "internals": "Market Internals",
}

_PROCESS_LABEL: dict[ProcessTier, str] = {
    "not_aligned": "Not aligned",
    "developing": "Developing",
    "near_ready": "Near ready",
    "actionable": "Strong",
}

_PHASE_LABEL: dict[SetupPhase, str] = {
    "early": "Early",
    "expansion": "Expansion",
    "extended": "Extended",
    "exhaustion": "Exhaustion",
}

_TRADEABILITY_LABEL: dict[TradeabilityBand, str] = {
    "strong": "Strong entry timing",
    "moderate": "Moderate entry timing",
    "weak": "Weak entry timing",
}


@dataclass(frozen=True)
class TradeabilityFlag:
    id: str
    label: str
    severity: FlagSeverity

    def to_dict(self) -> dict[str, str]:
        return {"id": self.id, "label": self.label, "severity": self.severity}


def _count_aligned(layers: list[dict[str, Any]], *, composite_bias: str) -> int:
    bias = composite_bias_from_summary(composite_bias)
    n = 0
    for lid in MATURATION_LAYER_KEYS:
        row = next((r for r in layers if str(r.get("layer") or "").lower() == lid), None)
        if row and _layer_aligned(row, composite_bias=bias):
            n += 1
    return n


def _process_tier(aligned: int, total: int = 6) -> ProcessTier:
    if aligned >= 5:
        return "actionable"
    if aligned == 4:
        return "near_ready"
    if aligned >= 2:
        return "developing"
    return "not_aligned"


def _quality_score(aligned: int, alignment_ratio: float | None, total: int = 6) -> int:
    base = int(round(100.0 * float(aligned) / float(total)))
    if alignment_ratio is not None:
        ar = max(0.0, min(1.0, float(alignment_ratio)))
        base = int(round(0.72 * base + 0.28 * (ar * 100.0)))
    return max(0, min(100, base))


def _resolve_phase_swing(daily_rsi: float | None) -> SetupPhase | None:
    if daily_rsi is None:
        return None
    if daily_rsi >= 80:
        return "exhaustion"
    if daily_rsi >= 70:
        return "extended"
    if daily_rsi >= 60:
        return "expansion"
    return "early"


def _resolve_phase_day(rsi: float | None) -> SetupPhase | None:
    if rsi is None:
        return None
    if rsi >= 75:
        return "exhaustion"
    if rsi >= 65:
        return "extended"
    if rsi >= 55:
        return "expansion"
    return "early"


def _extension_flags_swing(
    *,
    last: float | None,
    sma50: float | None,
    daily_rsi: float | None,
    params_extension_pct: float = 15.0,
    params_rsi_overbought: float = 70.0,
) -> list[TradeabilityFlag]:
    flags: list[TradeabilityFlag] = []
    if daily_rsi is not None:
        if daily_rsi >= 80:
            flags.append(
                TradeabilityFlag("rsi_exhaustion", f"RSI {daily_rsi:.0f} — exhaustion zone", "block")
            )
        elif daily_rsi >= 76:
            flags.append(
                TradeabilityFlag("rsi_extended", f"RSI {daily_rsi:.0f} — extended momentum", "block")
            )
        elif daily_rsi >= params_rsi_overbought:
            flags.append(
                TradeabilityFlag("rsi_extended", f"RSI {daily_rsi:.0f} — extended momentum", "warn")
            )
    if last is not None and sma50 is not None and sma50 > 0:
        ext = (last - sma50) / sma50 * 100.0
        if ext >= params_extension_pct:
            flags.append(
                TradeabilityFlag(
                    "above_sma50",
                    f"Price {ext:.0f}% above SMA50 — stretched vs mean",
                    "block" if ext >= params_extension_pct * 1.25 else "warn",
                )
            )
    return flags


def _extension_flags_day(
    *,
    rsi: float | None,
    last: float | None,
    session_open: float | None,
    atr: float | None,
    params_rsi_overbought: float = 70.0,
) -> list[TradeabilityFlag]:
    flags: list[TradeabilityFlag] = []
    if rsi is not None:
        if rsi >= 75:
            flags.append(
                TradeabilityFlag("rsi_exhaustion", f"Session RSI {rsi:.0f} — exhaustion zone", "block")
            )
        elif rsi >= params_rsi_overbought:
            flags.append(
                TradeabilityFlag("rsi_extended", f"Session RSI {rsi:.0f} — extended", "warn")
            )
    if last is not None and session_open is not None and atr is not None and atr > 0:
        move = abs(last - session_open)
        ratio = move / atr
        if ratio >= 2.0:
            flags.append(
                TradeabilityFlag(
                    "session_move_2x_atr",
                    f"Session move ~{ratio:.1f}× ATR — late for fresh entry",
                    "block",
                )
            )
        elif ratio >= 1.5:
            flags.append(
                TradeabilityFlag(
                    "session_move_1_5x_atr",
                    f"Session move ~{ratio:.1f}× ATR — pace already extended",
                    "warn",
                )
            )
    return flags


def _tradeability_score(flags: list[TradeabilityFlag], phase: SetupPhase | None) -> int:
    score = 100
    for f in flags:
        score -= 25 if f.severity == "block" else 12
    if phase == "extended":
        score -= 10
    elif phase == "exhaustion":
        score -= 20
    elif phase == "early":
        score += 5
    return max(0, min(100, score))


def _tradeability_band(score: int, flags: list[TradeabilityFlag]) -> TradeabilityBand:
    if any(f.severity == "block" for f in flags) or score < 40:
        return "weak"
    if score >= 70:
        return "strong"
    return "moderate"


def _primary_blocker(
    missing_layers: list[str],
    flags: list[TradeabilityFlag],
) -> str | None:
    blocks = [f for f in flags if f.severity == "block"]
    if blocks:
        return blocks[0].label
    if missing_layers:
        labels = [_LAYER_LABEL.get(l, l.title()) for l in missing_layers[:3]]
        return f"Missing alignment: {', '.join(labels)}"
    warns = [f for f in flags if f.severity == "warn"]
    if warns:
        return warns[0].label
    return None


def _watch_for(
    unlock_forecast: list[dict[str, Any]] | None,
    missing_layers: list[str],
    flags: list[TradeabilityFlag],
) -> str | None:
    if unlock_forecast:
        primary = next((h for h in unlock_forecast if h.get("is_primary_blocker")), None)
        if primary is None and unlock_forecast:
            primary = unlock_forecast[0]
        if primary:
            trigger = str(primary.get("trigger_condition") or "").strip()
            if trigger:
                return trigger
            dist = str(primary.get("distance_description") or "").strip()
            if dist:
                return dist
    if flags:
        block = next((f for f in flags if f.severity == "block"), None)
        if block:
            return f"Watch for pullback or structure reset — {block.label.lower()}"
    if missing_layers:
        lid = missing_layers[0]
        return f"Watch {_LAYER_LABEL.get(lid, lid)} for alignment with setup bias"
    return None


def build_setup_judgment(
    *,
    mode: Literal["day", "swing"],
    layers: list[dict[str, Any]],
    signal_summary: str,
    alignment_ratio: float | None = None,
    unlock_forecast: list[dict[str, Any]] | None = None,
    tech_result: Any | None = None,
    bars: list[Any] | None = None,
) -> dict[str, Any]:
    """Build API ``setup_judgment`` object for composite responses."""
    total = len(MATURATION_LAYER_KEYS)
    summary = (signal_summary or "neutral").strip().lower()
    binary_aligned = _count_aligned(layers, composite_bias=summary)
    # Setup progress uses directional agreement (verdict matches bias), not weighted ratio.
    aligned = binary_aligned
    tier = _process_tier(aligned, total)
    missing = derive_missing_layers(layers, composite_bias=composite_bias_from_summary(summary))

    phase: SetupPhase | None = None
    flags: list[TradeabilityFlag] = []

    if mode == "swing" and tech_result is not None:
        rsi = getattr(tech_result, "daily_rsi", None)
        sma50 = getattr(tech_result, "sma50", None)
        last = None
        if bars:
            try:
                closes = [float(b.close) for b in bars if getattr(b, "close", None)]
                last = closes[-1] if closes else None
            except (TypeError, ValueError):
                last = None
        phase = _resolve_phase_swing(float(rsi) if rsi is not None else None)
        flags = _extension_flags_swing(
            last=last,
            sma50=float(sma50) if sma50 is not None else None,
            daily_rsi=float(rsi) if rsi is not None else None,
        )
    elif mode == "day" and tech_result is not None:
        rsi = getattr(tech_result, "rsi", None)
        atr = getattr(tech_result, "atr", None)
        last = None
        session_open = None
        if bars:
            try:
                closes = [float(b.close) for b in bars if getattr(b, "close", None)]
                if closes:
                    last = closes[-1]
                    session_open = closes[0]
            except (TypeError, ValueError):
                pass
        phase = _resolve_phase_day(float(rsi) if rsi is not None else None)
        flags = _extension_flags_day(
            rsi=float(rsi) if rsi is not None else None,
            last=last,
            session_open=session_open,
            atr=float(atr) if atr is not None else None,
        )

    q_score = _quality_score(binary_aligned, alignment_ratio, total)
    t_score = _tradeability_score(flags, phase)
    t_band = _tradeability_band(t_score, flags)
    blocker = _primary_blocker(missing, flags)
    watch = _watch_for(unlock_forecast, missing, flags)

    return {
        "process": {
            "tier": tier,
            "label": _PROCESS_LABEL[tier],
            "layers_aligned": aligned,
            "layers_total": total,
        },
        "setup_phase": (
            {"id": phase, "label": _PHASE_LABEL[phase]} if phase is not None else None
        ),
        "tradeability": {
            "band": t_band,
            "label": _TRADEABILITY_LABEL[t_band],
            "flags": [f.to_dict() for f in flags],
        },
        "primary_blocker": blocker,
        "watch_for": watch,
        "engine_scores": {
            "quality": q_score,
            "tradeability": t_score,
        },
    }

File: frontend/lib/signal-evidence/setup-judgment.ts
/**
 * Setup judgment — quality (layer agreement) vs tradeability (entry geometry).
 * Parsed from composite `setup_judgment` or derived client-side from layers + technical row.
 */

import {
  resolveAlignmentDisplayTier,
  type AlignmentDisplayTier
} from "@/lib/alignment-display-tier";
import {
  layerRowEligibleForAlignmentCount,
  parseCompositeDirectionFields,
  resolveSignalsLayerAlignment,
  type SignalsLayerRowInput,
  type SignalsSetupBias
} from "@/lib/signals-page-present";

export type SetupPhaseId = "early" | "expansion" | "extended" | "exhaustion";
export type TradeabilityBand = "strong" | "moderate" | "weak";
export type TradeabilityFlagSeverity = "warn" | "block";

export type TradeabilityFlag = {
  id: string;
  label: string;
  severity: TradeabilityFlagSeverity;
};

export type SetupJudgment = {
  process: {
    tier: AlignmentDisplayTier;
    label: string;
    layersAligned: number;
    layersTotal: number;
  };
  setupPhase: { id: SetupPhaseId; label: string } | null;
  tradeability: {
    band: TradeabilityBand;
    label: string;
    flags: TradeabilityFlag[];
  };
  primaryBlocker: string | null;
  watchFor: string | null;
  /** Internal — not for hero UI. */
  engineScores?: { quality: number; tradeability: number };
};

const PROCESS_LABEL: Record<AlignmentDisplayTier, string> = {
  not_aligned: "Not aligned",
  developing: "Developing",
  near_ready: "Near ready",
  actionable: "Strong",
  invalidated: "Invalidated",
  re_evaluating: "Re-evaluating"
};

function reconcileSetupJudgmentProcess(
  judgment: SetupJudgment,
  composite: Record<string, unknown> | null | undefined,
  input: {
    bias: SignalsSetupBias;
    rows: SignalsLayerRowInput[];
    alignmentRatio?: number | null;
  }
): SetupJudgment {
  const dir = parseCompositeDirectionFields(composite ?? undefined);
  const biasAlignment = resolveSignalsLayerAlignment({
    rows: input.rows,
    bias: input.bias,
    alignmentRatio: input.alignmentRatio ?? null,
    compositeDirection: dir
  });
  const layersAligned =
    input.bias === "Neutral" && dir != null ? dir.directional : biasAlignment.aligned;
  const layersTotal = biasAlignment.total;
  const overstatedNeutral =
    input.bias === "Neutral" && judgment.process.layersAligned > layersAligned + 1;
  const drifted =
    Math.abs(judgment.process.layersAligned - layersAligned) > 1 ||
    judgment.process.layersTotal !== layersTotal;
  if (!overstatedNeutral && !drifted) return judgment;

  const tier = resolveAlignmentDisplayTier({
    layersAligned,
    layersTotal
  });
  return {
    ...judgment,
    process: {
      tier,
      label: PROCESS_LABEL[tier],
      layersAligned,
      layersTotal
    }
  };
}

const PHASE_LABEL: Record<SetupPhaseId, string> = {
  early: "Early",
  expansion: "Expansion",
  extended: "Extended",
  exhaustion: "Exhaustion"
};

const TRADEABILITY_LABEL: Record<TradeabilityBand, string> = {
  strong: "Strong entry timing",
  moderate: "Moderate entry timing",
  weak: "Weak entry timing"
};

function parsePhase(raw: unknown): { id: SetupPhaseId; label: string } | null {
  if (!raw || typeof raw !== "object") return null;
  const id = String((raw as { id?: string }).id || "").trim() as SetupPhaseId;
  if (!["early", "expansion", "extended", "exhaustion"].includes(id)) return null;
  const label = String((raw as { label?: string }).label || PHASE_LABEL[id]);
  return { id, label };
}

function parseFlags(raw: unknown): TradeabilityFlag[] {
  if (!Array.isArray(raw)) return [];
  const out: TradeabilityFlag[] = [];
  for (const row of raw) {
    if (!row || typeof row !== "object") continue;
    const id = String((row as { id?: string }).id || "").trim();
    const label = String((row as { label?: string }).label || "").trim();
    const sev = String((row as { severity?: string }).severity || "warn");
    if (!id || !label) continue;
    out.push({
      id,
      label,
      severity: sev === "block" ? "block" : "warn"
    });
  }
  return out;
}

export function parseSetupJudgment(body: unknown): SetupJudgment | null {
  if (!body || typeof body !== "object") return null;
  const root = (body as { setup_judgment?: unknown }).setup_judgment ?? body;
  if (!root || typeof root !== "object") return null;
  const proc = (root as { process?: unknown }).process;
  if (!proc || typeof proc !== "object") return null;
  const aligned = Number((proc as { layers_aligned?: number }).layers_aligned);
  const total = Number((proc as { layers_total?: number }).layers_total) || 6;
  if (!Number.isFinite(aligned)) return null;
  const tierRaw = String((proc as { tier?: string }).tier || "").trim();
  const tier = resolveAlignmentDisplayTier({
    layersAligned: aligned,
    layersTotal: total,
    maturationState: tierRaw === "actionable" ? "actionable" : null
  });
  const trade = (root as { tradeability?: unknown }).tradeability;
  const band = String((trade as { band?: string } | undefined)?.band || "moderate") as TradeabilityBand;
  const safeBand: TradeabilityBand =
    band === "strong" || band === "weak" || band === "moderate" ? band : "moderate";
  const engine = (root as { engine_scores?: unknown }).engine_scores;
  let engineScores: SetupJudgment["engineScores"];
  if (engine && typeof engine === "object") {
    const q = Number((engine as { quality?: number }).quality);
    const t = Number((engine as { tradeability?: number }).tradeability);
    if (Number.isFinite(q) && Number.isFinite(t)) {
      engineScores = { quality: q, tradeability: t };
    }
  }
  return {
    process: {
      tier,
      label: String((proc as { label?: string }).label || PROCESS_LABEL[tier]),
      layersAligned: Math.round(aligned),
      layersTotal: total
    },
    setupPhase: parsePhase((root as { setup_phase?: unknown }).setup_phase),
    tradeability: {
      band: safeBand,
      label: String((trade as { label?: string } | undefined)?.label || TRADEABILITY_LABEL[safeBand]),
      flags: parseFlags((trade as { flags?: unknown } | undefined)?.flags)
    },
    primaryBlocker: String((root as { primary_blocker?: string }).primary_blocker || "").trim() || null,
    watchFor: String((root as { watch_for?: string }).watch_for || "").trim() || null,
    engineScores
  };
}

function layerSupportsBias(row: SignalsLayerRowInput, bias: SignalsSetupBias): boolean {
  if (!layerRowEligibleForAlignmentCount(row)) return false;
  if (bias === "Bullish") return row.status === "Bullish";
  if (bias === "Bearish") return row.status === "Bearish";
  return row.status === "Bullish" || row.status === "Bearish";
}

function derivePhaseFromTechnicalReasoning(reasoning: string, mode: "swing" | "day"): SetupPhaseId | null {
  const m = reasoning.match(/RSI\s+(\d+(?:\.\d+)?)/i);
  if (!m) return null;
  const rsi = Number(m[1]);
  if (!Number.isFinite(rsi)) return null;
  if (mode === "swing") {
    if (rsi >= 80) return "exhaustion";
    if (rsi >= 70) return "extended";
    if (rsi >= 60) return "expansion";
    return "early";
  }
  if (rsi >= 75) return "exhaustion";
  if (rsi >= 65) return "extended";
  if (rsi >= 55) return "expansion";
  return "early";
}

function deriveFlagsFromTechnical(
  reasoning: string,
  mode: "swing" | "day"
): TradeabilityFlag[] {
  const flags: TradeabilityFlag[] = [];
  const rsiM = reasoning.match(/RSI\s+(\d+(?:\.\d+)?)/i);
  if (rsiM) {
    const rsi = Number(rsiM[1]);
    if (Number.isFinite(rsi)) {
      if (mode === "swing" && rsi >= 80) {
        flags.push({ id: "rsi_exhaustion", label: `RSI ${Math.round(rsi)} — exhaustion zone`, severity: "block" });
      } else if ((mode === "swing" && rsi >= 76) || (mode === "day" && rsi >= 75)) {
        flags.push({ id: "rsi_extended", label: `RSI ${Math.round(rsi)} — extended`, severity: "block" });
      } else if (rsi >= 70) {
        flags.push({ id: "rsi_extended", label: `RSI ${Math.round(rsi)} — extended momentum`, severity: "warn" });
      }
    }
  }
  const extM = reasoning.match(/(\d+(?:\.\d+)?)%\s+above\s+SMA50/i);
  if (extM) {
    const pct = Number(extM[1]);
    if (pct >= 15) {
      flags.push({
        id: "above_sma50",
        label: `Price ${pct.toFixed(0)}% above SMA50 — stretched vs mean`,
        severity: pct >= 19 ? "block" : "warn"
      });
    }
  }
  const atrM = reasoning.match(/~(\d+(?:\.\d+)?)×\s*ATR/i);
  if (atrM && Number(atrM[1]) >= 2) {
    flags.push({
      id: "session_move_2x_atr",
      label: `Session move ~${atrM[1]}× ATR — late for fresh entry`,
      severity: "block"
    });
  }
  return flags;
}

function tradeabilityScore(flags: TradeabilityFlag[], phase: SetupPhaseId | null): number {
  let score = 100;
  for (const f of flags) {
    score -= f.severity === "block" ? 25 : 12;
  }
  if (phase === "extended") score -= 10;
  else if (phase === "exhaustion") score -= 20;
  else if (phase === "early") score += 5;
  return Math.max(0, Math.min(100, score));
}

function tradeabilityBand(score: number, flags: TradeabilityFlag[]): TradeabilityBand {
  if (flags.some((f) => f.severity === "block") || score < 40) return "weak";
  if (score >= 70) return "strong";
  return "moderate";
}

/** Client fallback when API has not yet returned `setup_judgment`. */
export function deriveSetupJudgment(input: {
  mode: "swing" | "day";
  rows: SignalsLayerRowInput[];
  bias: SignalsSetupBias;
  alignmentRatio?: number | null;
  technicalReasoning?: string | null;
  unlockWatchFor?: string | null;
}): SetupJudgment {
  const alignment = resolveSignalsLayerAlignment({
    rows: input.rows,
    bias: input.bias,
    alignmentRatio: input.alignmentRatio ?? null
  });
  const tier = resolveAlignmentDisplayTier({
    layersAligned: alignment.aligned,
    layersTotal: alignment.total
  });
  const missing = input.rows
    .filter((r) => layerRowEligibleForAlignmentCount(r) && !layerSupportsBias(r, input.bias))
    .map((r) => r.name);
  const reasoning = (input.technicalReasoning || "").trim();
  const phase = reasoning ? derivePhaseFromTechnicalReasoning(reasoning, input.mode) : null;
  const flags = reasoning ? deriveFlagsFromTechnical(reasoning, input.mode) : [];
  const tScore = tradeabilityScore(flags, phase);
  const band = tradeabilityBand(tScore, flags);
  const primaryBlocker =
    flags.find((f) => f.severity === "block")?.label ??
    (missing.length > 0 ? `Key checks still disagree: ${missing.slice(0, 3).join(", ")}` : null) ??
    flags.find((f) => f.severity === "warn")?.label ??
    null;
  const watchFor =
    input.unlockWatchFor?.trim() ||
    (flags.find((f) => f.severity === "block")
      ? `What must change next: ${flags.find((f) => f.severity === "block")!.label.toLowerCase()}`
      : missing.length > 0
        ? `What must change next: ${missing[0]} needs to align with setup bias`
        : null);
  const qualityBase = Math.round((100 * alignment.aligned) / Math.max(1, alignment.total));
  const ar =
    input.alignmentRatio != null && Number.isFinite(input.alignmentRatio)
      ? Math.round(Math.max(0, Math.min(1, input.alignmentRatio)) * 100)
      : null;
  const quality = ar != null ? Math.round(0.72 * qualityBase + 0.28 * ar) : qualityBase;

  return {
    process: {
      tier,
      label: PROCESS_LABEL[tier],
      layersAligned: alignment.aligned,
      layersTotal: alignment.total
    },
    setupPhase: phase ? { id: phase, label: PHASE_LABEL[phase] } : null,
    tradeability: {
      band,
      label: TRADEABILITY_LABEL[band],
      flags
    },
    primaryBlocker,
    watchFor,
    engineScores: { quality, tradeability: tScore }
  };
}

export function resolveSetupJudgmentFromComposite(
  composite: Record<string, unknown> | null | undefined,
  input: {
    mode: "swing" | "day";
    rows: SignalsLayerRowInput[];
    bias: SignalsSetupBias;
    alignmentRatio?: number | null;
  }
): SetupJudgment | null {
  const parsed = parseSetupJudgment(composite);
  const layers = Array.isArray(composite?.layers) ? (composite!.layers as Record<string, unknown>[]) : [];
  const tech = layers.find((r) => String(r.layer || "").toLowerCase() === "technical");
  const reasoning = tech ? String(tech.reasoning || "") : "";
  const unlock = Array.isArray(composite?.unlock_forecast)
    ? String(
        (
          (composite!.unlock_forecast as Record<string, unknown>[]).find(
            (h) => h.is_primary_blocker
          ) ?? (composite!.unlock_forecast as Record<string, unknown>[])[0]
        )?.trigger_condition || ""
      )
    : null;
  const derived =
    parsed ??
    deriveSetupJudgment({
      mode: input.mode,
      rows: input.rows,
      bias: input.bias,
      alignmentRatio: input.alignmentRatio ?? null,
      technicalReasoning: reasoning,
      unlockWatchFor: unlock || null
    });
  return reconcileSetupJudgmentProcess(derived, composite ?? undefined, input);
}

/** Layer dots for list ranking (e.g. scanner). */
export function formatLayerProgressDots(aligned: number, total: number): string {
  const a = Math.max(0, Math.min(total, aligned));
  return `${"●".repeat(a)}${"○".repeat(Math.max(0, total - a))}`;
}

export function qualityBandFromAligned(aligned: number, total: number): "high" | "medium" | "low" {
  const pct = aligned / Math.max(1, total);
  if (pct >= 5 / 6) return "high";
  if (pct >= 2 / 3) return "medium";
  return "low";
}
